feat: add 'showcmdloc' option and statusline item

luukvbaal opened the pull request:
**vim-patch:9.0.1061: cannot display 'showcmd' somewhere else**

Problem:    Cannot display 'showcmd' somewhere else.
Solution:   Add the 'showcmdloc' option. (Luuk van Baal, closes vim/vim#11684)

https://github.com/vim/vim/commit/ba936f6f4e85cc1408bc3967f9fd7665d948909b

luukvbaal commented:
> `searchcount` is the other big one.

I'm not sure how worthwhile adding this one will be. It is relatively easy to write your own statusline component with `searchcount()`, leaving the presentation up to the user, which would be static when added as a builtin statusline item.

> Other items that come to mind are those mentioned in :h 'shortmess'

Thanks, I know I've missed the macro recording messages from that list at least.

> as well as the operator-pending message (the da shown before you complete with w).

This is covered by `'showcmd'`(and thus the current PR) right?

clason commented:
You're right; I was thinking of `reg_recording` etc.

luukvbaal commented:
I tried to write a test for this btw but it results in an empty statusline for some reason:
```vim
  " %S: 'showcmd' content, for use with 'cmdheight' = 0.
  set cmdheight=0
  set statusline=%S
  call assert_match('^\s*$', ScreenLines(&lines, &columns)[0])
  call feedkeys('1234')
  call assert_match('^\[1234\]s*$', ScreenLines(&lines, &columns)[0])
  set cmdheight&
  " result:
  " RunTheTest[44]..Test_statusline line 130: Pattern '^\\[1234\\]s*$' does not match '\[  occurs 80 times]'
```

Same thing with a lua screen test.

justinmk commented on src/nvim/option_defs.h:
```diff
@@ -336,6 +336,7 @@ enum {
   STL_ALTPERCENT      = 'P',  ///< Percentage as TOP BOT ALL or NN%.
   STL_ARGLISTSTAT     = 'a',  ///< Argument list status as (x of y).
   STL_PAGENUM         = 'N',  ///< Page number (when printing).
+  STL_SHOWCMD         = 'S',  ///< Showcmd buffer.
```
I wonder if dot `.` is semantically closer to what Vim would choose here. Not a blocker though.

luukvbaal replied:
I'm not sure I get what you mean. You're suggesting `.` instead of `S`?

I did come across some vi stackexchange questions to embed `showcmd` into the [statusline](https://vi.stackexchange.com/questions/2663/embed-showcmd-in-airline) and even [tabline](https://vi.stackexchange.com/questions/16857/showcmd-on-first-line-instead-of-last-line). Perhaps this change has value outside of `'cmdheight' == 0` context. So if we want to get Bram's feedback I guess I could check if he is open to including this(decoupled from `cmdheight`)?

justinmk replied:
> You're suggesting `.` instead of `S`?

yes



> So if we want to get Bram's feedback I guess I could check if he is open to including this(decoupled from `cmdheight`)?

if you're up for it, that's always a worth a check.

luukvbaal commented:
I haven't had a definitive answer if Bram is going to include the patch https://github.com/vim/vim/pull/11684 but his feedback seems to indicate that he will. Also awaiting answer to [this](https://github.com/vim/vim/pull/11684#issuecomment-1351659778) question:

> I wonder if adding a builtin `showcmd()` function, analogous to e.g. `mode()` and `searchcount()` will be more useful/idiomatic than adding a statusline item?

Any thoughts about it here? The general shape of the patch would stay the same but instead of a new statusline item we would have a builtin function that returns the showcmd buffer.

luukvbaal committed:
vim-patch:9.0.1061: cannot display 'showcmd' somewhere else

Problem:    Cannot display 'showcmd' somewhere else.
Solution:   Add the 'showcmdloc' option. (Luuk van Baal, closes vim/vim#11684)

https://github.com/vim/vim/commit/ba936f6f4e85cc1408bc3967f9fd7665d948909b

Co-authored-by: Luuk van Baal <luukvbaal@gmail.com>

luukvbaal commented:
Awaiting https://github.com/vim/vim/pull/11708, unless you want to merge this while that is pending.
